### ISSUE TITLE
Path tracing renders on Low quality texture

### ISSUE THREAD

**Issue by Drosaca** [gkjohnson/three-gpu-pathtracer#678]
I loaded a [3D model](https://sketchfab.com/3d-models/starlink-spacex-satellite-0a60f6720c5141c9a1c6d71aac108b31) and applied OrbitControls to it. The path tracing work fine when the model is at the origin (0,0,0). However, after repositioning the model far from the origin and moving the camera to the new position, the path tracing is rendered on a low-quality texture, while the default renderer displays the correct texture.
![chrome_raUJgxWO8g (1)](https://github.com/user-attachments/assets/d70a0526-97c2-4ebb-b47e-65e3d479e6e3)

**Comment by AntonPalmqvist** [comment]
Try increasing the textureSize parameter: https://github.com/gkjohnson/three-gpu-pathtracer?tab=readme-ov-file#texturesize

**Comment by gkjohnson** [comment]
When submitting an issue please follow the template and submit the information that's requested - you haven't provided enough information to say exactly what the issue is.

My guess is that this is related to floating point precision at large numbers away from zero (though you haven't said how far). This is something that's not currently handled internally so assuming that's what the issue is you'll have to adjust the model position and camera to be near zero.

**Comment by Drosaca** [comment]
> When submitting an issue please follow the template and submit the information that's requested - you haven't provided enough information to say exactly what the issue is.
> 
> My guess is that this is related to floating point precision at large numbers away from zero (though you haven't said how far). This is something that's not currently handled internally so assuming that's what the issue is you'll have to adjust the model position and camera to be near zero.

Ok I updated the issue, the far away value is "x": -3761603.743620786, "y": 4105279.00727716, "z": 3870314.8490743843. it looks like a low poly version of my object when it's far

**Comment by Drosaca** [comment]
> Try increasing the textureSize parameter: https://github.com/gkjohnson/three-gpu-pathtracer?tab=readme-ov-file#texturesize

i tried with 4096 but it doesn't change anything :(

**Comment by gkjohnson** [comment]
> Ok I updated the issue, the far away value is "x": -3761603.743620786, "y": 4105279.00727716, "z": 3870314.8490743843. it looks like a low poly version of my object when it's far but only with the path traced version.

Yes these numbers are very large and will likely cause numeric precision issues when rendering. I recommend moving the camera and model closer to the origin to address the issue. This could probably be fixed automatically by the renderer but it would need to be implemented.

> I tried with 4096 but it doesn't change anything :(

Yeah this won't help in this case, unfortunately, since it's a numeric problem.

**Comment by TheBlek** [comment]
I recently read a chapter from Ray Tracing Gems 2 [here](https://link.springer.com/book/10.1007/978-1-4842-7185-8) that seem to address this exact issue. In Chapter 34 Ingo Wald suggests moving ray origin along the direction closer to the intersected primitive and perform test. The resulting distance should be modified afterwards to reflect the origin displacement. With regards to choosing the distance to move he writes following:

> Even more generally, if one does have access to the bounding volume
hierarchy traverser’s distance to the leaf node that contained the primitive (or
if one used any sort of early-out bounding box test), then the distance to this
bounding volume is an excellent choice. 

With this, it sounds like the fix should be implemented in the three-mesh-bvh shader traversal step.
